File: src/main/scala/kim/dencka/scala/medium/BinaryTreeRightSideView.scala
package kim.dencka.scala.medium

object BinaryTreeRightSideView extends App {

  class TreeNode(
    val value: Int,
    val left: TreeNode = null,
    val right: TreeNode = null
  )

  val bt1 =
    new TreeNode(
      1,
      new TreeNode(
        2,
        new TreeNode(4, new TreeNode(8), new TreeNode(9)),
        new TreeNode(5, new TreeNode(10), new TreeNode(11))
      ),
      new TreeNode(
        3,
        new TreeNode(6, new TreeNode(12), new TreeNode(13)),
        new TreeNode(7, new TreeNode(14), new TreeNode(15))
      )
    )

  val bt2 =
    new TreeNode(
      1,
      new TreeNode(
        2,
        new TreeNode(
          3,
          new TreeNode(
            4,
            new TreeNode(
              5,
              new TreeNode(
                6,
                new TreeNode(8),
                new TreeNode(9)
              ),
              new TreeNode(7)
            )
          )
        )
      )
    )

  // KIM_INFO NOT WORKING
  def rightSideView(root: TreeNode): List[Int] = {
    def run(tree: TreeNode, isRight: Boolean): List[Int] =
      if (tree == null) List()
      else if (tree.right != null)
        if (isRight) List(tree.value) ++ run(tree.right, isRight = true)
        else run(tree.right, isRight = true)
      else if (tree.right != null && tree.left != null)
        if (isRight) List(tree.value) ++ run(tree.right, isRight = true) ++ run(tree.left, isRight = false)
        else run(tree.right, isRight = false) ++ run(tree.left, isRight = false)
      else if (tree.right == null) List(tree.value) ++ run(tree.left, isRight = true)
      else if (isRight) List(tree.value) ++ run(tree.right, isRight = true) ++ run(tree.left, isRight = false)
      else run(tree.right, isRight = true) ++ run(tree.left, isRight = false)

    run(root, isRight = true)
  }

  println(rightSideView(bt2))
  // expected [1,2,3,4,5,7,9]
  // reality [[1,2,3,4,5,7]

}

class BinaryTreeRightSideView

File: src/main/scala/kim/dencka/java/medium/BinaryTreeRightSideView.java
package kim.dencka.java.medium;

import kim.dencka.java.payloads.TreeNode;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.List;
import java.util.Queue;

public class BinaryTreeRightSideView {
    public static void main(String[] args) {
        TreeNode tree = TreeNode.fromArray(new Integer[]{1, 2, 3, 4, null, null, null, 5});
        System.out.println(rightSideView(tree)); // [1, 3, 4, 5]
    }

    // BFS
    public static List<Integer> rightSideView(TreeNode root) {
        if (root == null) return List.of();
        Queue<TreeNode> queue = new ArrayDeque<>();
        List<Integer> result = new ArrayList<>();

        queue.offer(root);
        while (!queue.isEmpty()) {
            int curSize = queue.size();
            for (int i = 0; i < curSize; i++) {
                TreeNode cur = queue.poll();
                if (i == curSize - 1) result.add(cur.val);
                if (cur.left != null) queue.offer(cur.left); // first left
                if (cur.right != null) queue.offer(cur.right); // then right
            }
        }

        return result;
    }

    // DFS
//    public static List<Integer> rightSideView(TreeNode root) {
//        List<Integer> result = new ArrayList<>();
//        dfs(root, result, 0);
//        return result;
//    }
//
//    public static void dfs(TreeNode tree, List<Integer> result, int level) {
//        if (tree == null) return;
//
//        // if this is our first time at this level, add the current node
//        if (level == result.size()) result.add(tree.val);
//
//        // first right, then left
//        dfs(tree.right, result, level + 1);
//        dfs(tree.left, result, level + 1);
//    }

}
